### DISCUSSION TITLE
How should we handle `timestamp` in v1.0?

### DISCUSSION THREAD

**Discussion by karlfloersch** [ethereum-optimism/optimistic-specs#23] [Q&A]
# Design Goals

1. Frequent updates to the timestamp.
2. Reliable timestamps - ie. the sequencer cannot manipulate the timestamp.
3. Accurate timestamps - ie. the timestamp does not lag behind the current time.

## Additional Context
In v1.0 we will be able to supply easy access to all information contained in recent L1 blocks. This is because every L1 block triggers a `deposit block` which updates: `lastL1Timestamp`, `lastL1BlockNumber`, `lastL1BaseFee`, and `lastL1BlockHash`. The sequencer will pull blocks trailing L1 by `CONFIRMATION_DEPTH` which we expect to be approx 5 minutes, possibly less. This would imply that the L1 timestamp is:

1. Updated frequently - updates ~ every 15 seconds
2. Reliable - the sequencer cannot manipulate it
3. Not accurate. The timestamp would lag behind the current time by whatever the confirmation depth is. 

Given these design goals, how should we handle `timestamp` in v1.0?

**Comment by karlfloersch** [comment]
So far there are two main proposals:

## Proposal 1: Add timestamp offset == to confirmation time
If the confirmation time is 5 minutes, then we would offset all L1 timestamps by 5 minutes. That would satisfy property (3) and make for a frequently updated, reliable, and accurate timestamp.

- Pros
  - Simple
  - Only one timestamp so devs don't get confused
- Cons
  - Updated every 15 seconds (is that too slow?)
  - Hides complexity from devs - why not give access to all info?

#### Possible amendment to this proposal
Expose `sequencerTimestamp` on-chain which is simply equal to `lastL1Timestamp+sequencerOffset`. Additionally, allow the sequencer to update `sequencerOffset` with a normal transaction on-chain. This means the sequencer **can** update their timestamp every block, but it is not mandatory at the protocol level. This gives you the same granularity as option (2) without enshrining the timestamp updates.

## Proposal 2: Allow sequencer to add their own timestamps
In addition to the `lastL1Timestamp` from each deposit block, the sequencer also submits a timestamp with every block. This timestamp has some constraints on it - eg. it can't be >5 minutes more than the `lastL1Timestamp`. 

- Pros
  - More granular timestamps (can be as granular as we want!)
  - Let's devs choose what timestamp they want to use (lastL1Timestamp, sequencer timestamp, etc)
- Cons
  - Fundamentally more calldata for every block (now you need to include a timestamp with your sequencer blocks)
  - More protocol complexity

**Reply by maurelian** [reply]
Thoughts regarding Proposal 1: 

I generally lean towards simplicity, so I think the non-amended version of proposal ! should be taken as the 'default' unless strong arguments compel us to add complexity.

> Additionally, allow the sequencer to update `sequencerOffset` with a normal transaction on-chain.

This would allow for the `sequencerTimestamp` to be non-monotonic, which I don't think would be very useful to developers. 

Thoughts regarding Proposal 2: 

From the app dev perspective, I think there might be some concern that this introduces ambiguity about the canonical source of time. Ie. what issues might be introduced if Alice's contract uses `lastL1Timestamp` for time, and Bobs contract uses `sequencerTimestamp`?

**Comment by maurelian** [comment]
Some basic questions about the requirements:

> Frequent updates to the timestamp

What does 'frequent' mean? Should it be **more frequent** than the L1 block time of ~15 seconds?

> This is because every L1 block triggers a deposit block which updates: `lastL1Timestamp`, `lastL1BlockNumber`, `lastL1BaseFee`, and `lastL1BlockHash`

How are these values accessed on L2? Are they opcodes, if so, how does that affect the goal of EVM equivalence? (The question also applies to `sequencerOffset` and 
`sequenceTimestamp` referenced in the proposals). 

> by `CONFIRMATION_DEPTH` which we expect to be approx 5 minutes. 

I keep hearing this statement, which feels like a mismatch of unit. This will be measured in blocks, not actually minutes right? If so, I think saying "approx 5 minutes of blocks" would be more precise.

**Reply by ben-chain** [reply]
>How are these values accessed on L2? Are they opcodes, if so, how does that affect the goal of EVM equivalence? 

I think the most EVM-equivalent proposal would be to make them accessible via calls to a predeploy, and updated by deposit transactions.  This would afford any EVM-level differences to be relegated to the new deposit type, which is a fundamental difference that we can't avoid.

>What does 'frequent' mean? Should it be more frequent than the L1 block time of ~15 seconds?

Personally I think this is ideal, but I will sync with Kevin on this who knows best

**Reply by ben-chain** [reply]
Synced with Kevin -- in his view, a frequency matching L1 is much better than a lower frequency, but a frequency exceeding L1 block times would be ideal.

**Reply by tynes** [reply]
Having higher frequency updates to the timestamp is something that applications would want, but it seems less safe to be incrementing the timestamp more frequently without having higher guarantees of finality.

**Comment by ben-chain** [comment]
## Notes on Reliability
I just want to add some color around the design goal of "reliable timestamps" -- I think it's worth more precisely defining what this means. In particular, there is a fundamental property of sequencing that cannot be solved by any of these proposals. I think that because this doesn't change, and is the most important timing assumption to consider for developers, this decision isn't best informed by reliability, but rather UX, consistency, and simplicity. I would state it simply as:

_"The sequencer may insert transactions up to `FORCE_INCLUSION_PERIOD` 'into the past'."_

The above statement is how to think of/word it "in L1 time."  Conversely, "in L2 time" (i.e. from an L2 smart contract's perspective), given some `lastL1Timestamp` value corresponding to the last "sequencer ingested" L1 block, then we must assume that a malicious sequencer could have generated the current transaction with full knowledge of the L1 blocks in the range `[lastL1Timestamp, lastL1Timestamp + FORCE_INCLUSION_PERIOD)`. Those are L1 blocks that are "in the future"/"not yet processed" from the perspective of L2, but the sequencer can see them at the time it generates those L2 batches.

This is a fundamental requirement of enabling sequencing. To allow sequencing to "lock in" a reliable L2 state ahead of time, the system needs a `FORCE_INCLUSION_PERIOD` roughly equivalent to the sum of 1) a safe L1 confirmation depth (guaranteeing the sequencer will only ingest deposits once they aren't going to re-org out of L1), and 2) a safe time-to-batch-submission (guaranteeing the sequencer has enough time to submit a batch after processing the deposits and any transactions).

This manifests in some timing restrictions that cannot be avoided by application developers.  The simplest one is that a commit-reveal scheme cannot have a reveal phase which starts <`FORCE_INCLUSION_PERIOD` after the commit phase ends, even if that period is longer than what could be chosen on L1 as a reorg-safe window. There is no proposal we could make here which changes that.

Given this fundamental restriction, the "reliability" that we *can* give to the L2 timestamps are to have *monotonicity*: timestamps can be required to strictly increase, between sequencer AND L2 transactions. Enforcing this has the following consequences that application developers can trust:
   - The current L2 transaction's timestamp is the greatest that it has ever been for all transactions so far.
   - The current L2 transaction's timestamp is not more than `FORCE_INCLUSION_PERIOD` in the past vs. the L1 block in which the batch was submitted.
   - The current L2 transaction's timestamp is not in greater than the L1 block in which the batch was submitted.

(note: In principle, we could also very slightly strengthen the last guarantee so that the current L2 timestamp is not greater than `time_of_l1_batch_confirmation - l1_confirmation_depth` which would basically argue that a sequencer that says their block timestamp is the same as their submission time is being malicious because at the time it was sent to the mempool it should've been lower.)

## Notes on Equivalence Between some of these options
As a note, the addendum to Proposal 1. basically makes it equivalent to Proposal 2.  This is because the "sequencer sets the timestamp" is equivalent to setting an offset of "whatever timestamp the sequencer wants" - "the actual L1 block timestamp".

**Comment by ben-chain** [comment]
Because of the fundamental security complexity in making L2 timing assumptions, and the desire to provide integrators with something simple, I'm currently an advocate of Proposal 2, with a monotonicity enforcement. I think the argument for this is:
1. **Does not give a false sense of security.** I would rather say to an integrator:
   >The sequencer assigns a timestamp dynamically, it its version of "real-time." In the worst case (sequencer goes down, or is malicious) this can cause a deposit's timestamp to increase by up to `sequencing_window`.  Therefore, do not trust time at a granularity exceeding `sequencing_window`.

   than say:
   > Well, there is this offset which gets assigned to deposits based on the confirmation depth, and that's based on L1. This is more resilient to a sequencer which goes offline or rogue.  BUT, the sequencer can still do these insertion operations; so really that is only resilient if you are doing X or Y, and not Z...

   IMO the first is both easier to communicate to integrators, and less likely to result in an application developer accidentally believing that they can trust the time, when in reality they shouldn't.
2. **Best-case accuracy**: the timestamps are maximally granular and minimally wrong in the happy case. They are also not dependent on the L1 timestamp being accurate, or on the L1 node receiving new L1 blocks consistently.

I think that the most likely counterarguments to this would be if:
- The complexity of enforcing monotonicity is way higher without an unmodified Proposal 1.--either by making the encoding/decoding extremely crazy, or by requiring some weird additional state to be tracked.
- The modifications that we have to make to EIP1559 become significantly more complex with this version of the protocol, either in terms of what the algorithm has to do, or how that algorithm impacts sequencer behavior during edge cases.

**Reply by karlfloersch** [reply]
You mention EIP1559, but I do want to call out that there is a chance that by giving the sequencer more control of the timestamp (eg. in proposal 2) it could make the EIP1559 calculation smoother for the sequencer. So specifically the challenge with EIP1559 calculations is that fundamentally L1 is the source of time in L2, but L1 is:

1. Slower than L2 (ie. L1 has a ~13 second block time vs super low blocktimes in L2)
2. Somewhat irregular because PoW does not provide a consistent block time 

Additionally, there may be cases where the sequencer disconnects from L1 very briefly and misses a block, and then shortly thereafter reconnects and syncs the missed block and the next block. In this case it would be ideal if, as long as the disconnection is short, the EIP1559 algorithm that we use does not get negatively effected. An example of a negative effect would be imagine you give the sequencer a fixed number of blocks that they can insert _per L1 block_, this means that if they get L1 blocks in quick succession (maybe they went offline), then they would be missing out on blocks that they can insert by applying both L1 blocks right after each other. This is a really tricky problem!

It seems to me the best solution that I can think of to this problem is to go with an approach like proposal 2 and use the **sequencer timestamp** as what is used when calculating the EIP1559 prices. Because the timestamps are bound by the protocol (eg. monotonicity is enforced, the sequencer can't set the timestamp to something way in the future or past, etc) this means that EIP1559 can still correctly meter the gas on L2. However, the sequencer also has enough freedom to in the honest case follow an EIP1559 protocol that accounts for very short term disconnects from L1 or just irregular block creation timing.

So given all this, I'm very much in favor of:

1. Force the sequencer to submit a timestamp for each block. This timestamp is used in EIP1559. This is called `sequencerSuggestedTimestamp`
2. We store the `lastL1Timestamp` for every deposit
3. We expose a function `getSequencerTimestamp()`:
```
def getSequencerTimestamp():
  if sequencerSuggestedTimestamp < lastL1Timestamp:
    return lastL1Timestamp
  if sequencerSuggestedTimestamp > lastL1Timestamp + sequencingWindow:
    return lastL1Timestamp + sequencingWindow
  return sequencerSuggestedTimestamp
```
4. We use `getSequencerTimestamp()` for the EIP1559 calculation.
5. We make the `TIMESTAMP` opcode return `getSequencerTimestamp()` as well (to make sure we only have 1 timestamp for @smartcontracts 😁)

cc @protolambda

**Comment by smartcontracts** [comment]
I need to read this thread in more detail, but I just want the following properties out of any proposal:

- There is a SINGLE timestamp.
- The timestamp can be trusted to be correct within some known bound.
- The timestamp is monotonic.

**Reply by ben-chain** [reply]
Yes, agreed.  My proposal would fit these properties, where the "known bound" is equal to either `sequencing_window` (aka FIP) or maybe `sequencing_window - expected_batch_confirmation_time`.

**Comment by tynes** [comment]
How would timestamps be handled in the case of a reorg larger than the `CONFIRMATION_DEPTH`? If reorgs on L2 are triggered by reorgs on L1 are developed as a first class citizen, then its possible that we could eliminate the concept of a `CONFIRMATION_DEPTH` on L2 and have the L1 timestamp be about equal to the L2 timestamp. That would increase the reward of reorging L1 for miners/validators, perhaps more than the consensus developers/researchers are considering.

**Reply by ben-chain** [reply]
>How would timestamps be handled in the case of a reorg larger than the `CONFIRMATION_DEPTH`?

Very good question. I think the right answer would be:
 - For the deposits which are already older than the `sequencing_window`, then the protocol would already have given them a timestamp, so that would be handled naturally by the sequencer re-syncing the new L1. 
 - From there, I guess you'd need logic to make sure that you dealt with any deposits which were about to expire (i.e. not yet finalized, but WILL be finalized before the sequencer will have time to submit the next batch) and pre-empt what timestamp the protocol will give them when they do expire.
 - From there, the sequencer would return to setting the timestamps to `Date.now()`.

 So basically, the sequencer would observe what deposits have already expired or are just about to (and therefore will been given a timestamp by the protocol, outside of its control) and just use that.  Then, once it gets to some deposits which are recent enough for it to set the time for, it returns to using `Date.now()`. This recovery definitely introduces some complexity, hopefully manageably, but it's a great callout.

>its possible that we could eliminate the concept of confirmation depth

I think that limiting the frequency of L2 reorgs is a very worthwhile design goal.  We want the sequencer to provide a consistent view of the L2 state in advance of it "arriving" on L1--not one which will change every time there's an uncle. Eventually we want the sequencer to have a bond for upholding this pending state.  This means that you need to wait for the confirmation depth before ingesting deposits into L2.

**Reply by tynes** [reply]
I don't love the idea of using `Date.now` to set timestamps on L2 without being really sure that they cannot be manipulated by the infrastructure that the sequencer runs on. Clocks naturally [drift](https://en.wikipedia.org/wiki/Clock_drift) on computers over time - how much drift is acceptable? I'm imagining a situation where a cloud vm has its time manipulated without the sequencer realizing it. For this reason, I like using L1 based timestamps since they cannot be manipulated in this way.

**Reply by ben-chain** [reply]
CC @protolambda on this, I'm out of my league at this level. I'm handwaving `Date.now` as however L1 block producers set it, maybe there's something else to use or some external service.  But it's counterintuitive to me that this would be easier or result in a smaller diff, considering "use the current time, make sure monotonic" is roughly what L1 nodes do when producing blocks.

**Reply by ben-chain** [reply]
Being able to control the drift on our end also seems better to me than being forced to succumb to L1's drift, which is out of our control.

**Comment by NoahZinsmeister** [comment]
hi! just wanted to hop in with a few quick thoughts on the implications of different timestamp schemes for Uniswap oracles, as an example of some of the things an application developer might be thinking about.

first, timestamp monotonicity is taken for granted in the logic, so it's important to preserve that.

second, `block.timestamp` is hard-coded, so it'd be great if that's the API through which the value is exposed, and also ideally the only source of time, to encourage consistent cross-application usage.

third, from my perspective a timestamp update cadence of `CONFIRMATION_DEPTH` is too long for this setting. i imagine it would be fairly trivial to set the price on a pool to a manipulated value in the first tx after the timestamp update, backrun in the next tx to avoid arbitrage losses, and then reap the rewards of having manipulated this value for a whole ~5 minutes. note: this problem isn't even fully ameliorated by shorter update cadences, as the fundamental issue at play here is the 1 tx/block model, which may be out of scope for this conversation. ​i'm curious what the `sequencing_window` duration is expected to be, as that has implications as well.

**Reply by ben-chain** [reply]
In the context of uniswap oracle manipulation, I've done some thinking about this in the past and have a view.

I believe that how we expose time is actually _less_ important than how the sequence of TX inputs can be manipulated. In other words, the attack in "Notes on Reliability"--is fundamental, no matter how we display time--is where the uni oracle runs into issues.

Even with multiple transactions per block, the uniswap oracle on L1 relies on the assumption that *future* blocks are determined by *different* miners.  This is basically assuming that miners can't do multi-block backrunning. I believe that if miners could do time bandit/multi-block attacks on L1, the uni oracle would be vulnerable on L1 too.

That assumption is much much weaker on L2. With a single sequencer, it's a trivial attack, regardless of multi-tx blocks or single-tx blocks. With multiple sequencers, then you can do the oracle update whenever the current consensus leader changes, under the assumption that the previous and next BP are not colluding.

**Reply by ben-chain** [reply]
Here is the best "sequencer-resilient" uniswap oracle construction I've been able to come up with.

Basically, you rely on "force included" transactions (transactions from the censorship-resistant path) to counteract the sequencer's control. Because the sequencer has no control over the [deposit block generation function](https://github.com/ethereum-optimism/optimistic-specs/blob/main/components/rollup_node/block_gen.md#deposit-block-generation), we can try to use these blocks as the "un-backrunnable" ones. (the sequencer can manipulate *when* those deposit blocks get ingested, delaying them up to the `sequencing_window`, but not *what* those block contents are)

To do this successfully you must:
- only update the oracle when there is a deposit block -- think of this as the L2 equivalent of "the start of a new, non-controlled-by-previous-miner L1 block" which L1 uniswap relies on. 
- Assume or enforce that the sequencer cannot consistently become the first deposit in an L1 block (which would therefore make it the first transaction in the L2 deposit block, allowing for successful backrun).

The second part is tricky.  I think it's a reasonable assumption over many blocks in aggregate, but the problem is that for any block in which the sequencer *does* manage to be the first, then they can backrun for those blocks. Because they choose the sequence once the L1 blocks have already occurred, they can just wait until they *do* get to be the first transaction in a deposit, and selectively manipulate before those deposit blocks. In other words, it's a free option to try and get in first, and then free to manipulate only once successful. This also applies when you try naive solutions like "randomize deposit block order"--they're only partial solutions.

I think a robust solution would require a flashbots-style auction on deposit block ordering from within the L2 state.  With this, arbers could bid up to the arb opportunity's size to get their transaction first in the block, and it would truly become unprofitable for the sequencer in all cases.

Even without that robust solution, I think a combo of these approaches and a multi-sequencer protocol are the path towards a reasonably robust oracle on L2. Given the complexity, though, I suspect that passing L1 oracle updates into L2 is a very reasonable short-term path forward.

**Reply by ben-chain** [reply]
> ​i'm curious what the sequencing_window duration is expected to be, as that has implications as well.

10 minutes is usually what we throw around.  The precise answer is `deposit_confirmation_depth + batch_submission_ceil`, where `batch_submission_ceil` is the upper bound on how long it takes for the batch submitter to reliably roll up L2-confirmed transactions as L1 calldata.

**Reply by NoahZinsmeister** [reply]
thanks for the replies! i see where you're coming from - point taken that if our oracles are not robust to a malicious sequencer we've probably lost the plot...

i think i'm also convinced that the "force-included" transaction dynamic is the closest equivalent to the assumption that mining sequential blocks is hard on l1. re. your second requirement - my worry is that it would be a long time for efficient MEV markets to develop around this behavior, and while i can't say for sure, it feels to me like the uncertainty around ultimate l2 tx ordering + the medium-high latency of `sequencing_window` might either make atomic arb impossible or at least quite hard. perhaps i'm wrong about this though, maybe worth gaming out some scenarios.

> suspect that passing L1 oracle updates into L2 is a very reasonable short-term path forward.

agreed that this should probably be the "official" short-term recommendation

i'll sit on this a bit, it's an interesting design space.

**Reply by ben-chain** [reply]
>my worry is that it would be a long time for efficient MEV markets to develop around this behavior

Yeah, this is definitely another significant concern, especially because of this "wait and see" attack where the sequencer can make the arb fail when it sees someone actually trying. I think that cross-chain aggregation protocols which are actively bringing in big aggregated trades from other chains already would be the most likely first source of this, but it's not clear.

>it feels to me like the uncertainty around ultimate l2 tx ordering + the medium-high latency of `sequencing_window` might either make atomic arb impossible

I'm pretty confident it's *possible* to make the arb transactions atomic -- remember, you just use the pathway for *initiating* the transaction. You don't have to deposit in the funds with the transaction, you can use funds you already have in L2 and just trigger the trade via L1. These transactions can run whatever logic in L2 when triggered to atomically decide whether to arb. The uncertainty in final ordering is not an issue for the transaction's atomicity, but it *is* an issue for the trade ever actually meeting its atomic condition to execute.

**Reply by NoahZinsmeister** [reply]
hmm ok fair enough. seems like the happy path here is a mexican standoff where no one tries to manipulate because it would be punished/back-run/etc. by "force-includers" with appropriately flexible l2 logic (really struggling not to make a force includooooors meme)

anyway, this is definitely all dependent on efficient MEV markets, excited to see developments there!
